### PR TITLE
Add scenario deduplication validation with test_minimal_number_of_scenarios

### PR DESCRIPTION
**Pull request by jandro996**
## Motivation

As the system-tests repository grows, we need to ensure efficient test matrix management by preventing duplicate scenarios and identifying opportunities to merge similar EndToEnd scenarios. Duplicate scenarios waste CI resources and increase maintenance overhead without providing additional test coverage.

## Changes

Added scenario deduplication test (test_minimal_number_of_scenarios.py):

Three-level validation system:
- LEVEL 1: Uses comprehensive comparison of configuration parameters with _get_difference() function
- LEVEL 2: Validates identical environment variables (weblog_env and agent_env)
- LEVEL 3: Checks subset compatibility where one scenario's environment can be included in another

Test behavior:
- Fails the test when scenarios have equivalent configurations AND identical environment variables (mandatory merge)
- Also fails when scenarios have equivalent configurations with compatible weblog_env (subset relationship)
- Provides detailed logging of differences between scenarios

Configurable skip rules:
- SKIP_MERGE_SCENARIOS dictionary allows defining scenario pairs that cannot be merged due to specific test requirements
- Supports bidirectional restrictions with detailed reasoning
- Keeps an organized list of scenarios that should not be merged

Detailed comparison:
- Normalizes environment variables for accurate comparison
- Recursively compares objects to find differences
- Logs detailed information about differences between dictionaries

This test aims to maintain an optimal test matrix by identifying redundant scenarios while preserving necessary test coverage.

## Workflow

1. ⚠️ Create your PR as draft ⚠️
2. Work on you PR until the CI passes
3. Mark it as ready for review
    * Test logic is modified? -> Get a review from RFC owner.
    * Framework is modified, or non obvious usage of it -> get a review from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)

:rocket: Once your PR is reviewed and the CI green, you can merge it!

🛟 [#apm-shared-testing](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X) 🛟

## Reviewer checklist

* [ ] If PR title starts with `[<language>]`, double-check that only `<language>` is impacted by the change
* [ ] No system-tests internal is modified. Otherwise, I have the approval from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)
* [ ] A docker base image is modified?
    * [ ] the relevant `build-XXX-image` label is present
* [ ] A scenario is added (or removed)?
    * [ ] Get a review from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)

### CODE REVIEW

File: tests/test_the_test/test_minimal_number_of_scenarios.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        excluded_vars = {
+            'SYSTEMTESTS_SCENARIO',  # This is always different per scenario
+            'DD_LOGS_INJECTION',     # May be added dynamically
+        }
+        
+        normalized = {}
+        for key, value in env_dict.items():
+            if key not in excluded_vars:
+                # Convert None values to empty string for consistent comparison
+                normalized[key] = str(value) if value is not None else ""
```
Comment (cbeauchesne) [inline tests/test_the_test/test_minimal_number_of_scenarios.py:37]: Setting `None` in env removes the value if it has been defined, so I think we cannot convert None to `""`

Comment (cbeauchesne) [reply tests/test_the_test/test_minimal_number_of_scenarios.py:37]: FYI, Code to play with the SDK, to see how it behaves : 

```python
import docker

client = docker.from_env()
output = client.containers.run(
    "alpine",
    "env",
    environment={"HOSTNAME": None, "FOO": ""},  # removes hostname, declare foo as enpty string
    remove=True
)
print(output.decode())
```

File: tests/test_the_test/test_minimal_number_of_scenarios.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        
+        return normalized
+    
+    # Compare each scenario with every other scenario
+    for i, scenario_a in enumerate(endtoend_scenarios):
+        for j, scenario_b in enumerate(endtoend_scenarios):
+            if i >= j:  # Skip self-comparison and duplicate pairs
+                continue
```
Comment (cbeauchesne) [inline tests/test_the_test/test_minimal_number_of_scenarios.py:45]:
```
import itertools

for scenario_a, scenario_b in itertools.combinations(endtoend_scenarios, 2):
```

File: tests/test_the_test/test_minimal_number_of_scenarios.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+                
+            # Check if either scenario is marked to skip merge
+            if getattr(scenario_a, 'skip_merge', False):
```
Comment (cbeauchesne) [inline tests/test_the_test/test_minimal_number_of_scenarios.py:48]: Could you keep an exclusion list here, rather extending the scenario API ? 

```
excluded_couples = [
  (appsec_blocking_full_denylist, other_stuff),  # there is a good reason to not merge them
]

...

if (scenario_a, scenario_b) in excluded_couples:
   continue

File: tests/test_the_test/test_minimal_number_of_scenarios.py [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+from datetime import datetime, UTC
+from pathlib import Path
+
+import pytest
+
+from utils import scenarios, context
+from utils._context._scenarios import get_all_scenarios
+from utils._context._scenarios.endtoend import EndToEndScenario
+
+logger = logging.getLogger(__name__)
```
Comment (cbeauchesne) [inline tests/test_the_test/test_minimal_number_of_scenarios.py:13]: There is a `logger` inside `utils`

File: utils/_context/_scenarios/endtoend.py [hunk 5]
```diff
@@ -803,3 +822,90 @@ def get_junit_properties(self) -> dict[str, str]:
         result["dd_tags[systest.suite.context.appsec_rules_file]"] = self.weblog_container.appsec_rules_file or ""
 
         return result
+
+    def __eq__(self, other: object) -> bool:
+        """Compare EndToEndScenario instances excluding name, doc and weblog_env"""
+        if not isinstance(other, EndToEndScenario):
```
Comment (cbeauchesne) [inline utils/_context/_scenarios/endtoend.py:828]: Exclude inherited classes : 

```suggestion
        if type(other) is not EndToEndScenario:
```

File: tests/test_the_test/test_minimal_number_of_scenarios.py [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+                    "larger_scenario": larger_scenario,
+                    "reason": "equivalent configurations with compatible weblog_env (subset relationship)",
+                    "new_variables_to_insert": new_variables_to_insert,
+                    "smaller_weblog_env": smaller_env,
+                    "larger_weblog_env": larger_env,
+                }
+            )
+
+    # Generate report file with potential merges
+    if potential_merges:
```
Comment (cbeauchesne) [inline tests/test_the_test/test_minimal_number_of_scenarios.py:233]: I'd rather exit in error. Report in test_the_test will be completly ignored. WDYT ?
